src/main/java/org/apache/brooklyn/cloudfoundry/entity/CloudFoundryEntityImpl.java

```diff
@@ -0,0 +1,352 @@
+    protected void initEnrichers() {
+        super.initEnrichers();
+        ServiceStateLogic.ServiceNotUpLogic
+                .updateNotUpIndicator(this, SERVICE_PROCESS_IS_RUNNING,
+                        "No information yet on whether this service is running");
+    }
+
+    @Override
+    public abstract Class getDriverInterface();
+
+    @Override
+    public EntityPaasDriver getDriver() {
+        return driver;
+    }
+
+    @Override
+    public final void start(final Collection<? extends Location> locations) {
+        if (DynamicTasks.getTaskQueuingContext() != null) {
+            doStart(locations);
+        } else {
+            Task<?> task = Tasks.builder().name("start (sequential)").body(new Runnable() {
+                public void run() {
+                    doStart(locations);
+                }
+            }).build();
+            Entities.submit(this, task).getUnchecked();
+        }
+    }
+
+    protected final void doStart(Collection<? extends Location> locations) {
+        ServiceStateLogic.setExpectedState(this, Lifecycle.STARTING);
+        try {
+            preStart(findLocation(locations));
+            customStart();
+            log.info("Entity {} was started", new Object[]{this});
+            postDriverStart();
+
+            ServiceStateLogic.setExpectedState(this, Lifecycle.RUNNING);
+        } catch (Throwable t) {
+            ServiceStateLogic.setExpectedState(this, Lifecycle.ON_FIRE);
+            log.error("Error error starting entity {}", this);
+            throw Exceptions.propagate(t);
+        }
+    }
+
+    protected void preStart(Location location) {
+        this.addLocations(MutableList.of(location));
+        if (getLocationOrNull() != null) {
+            cfLocation = getLocationOrNull();
+        } else {
+            throw new ExceptionInInitializerError("Location should not be null in " + this +
+                    " the entity needs a initialized Location");
+        }
+        initDriver(cfLocation);
+    }
+
+    /*
+     * TODO: avoiding boilerplate code
+     * This method was gotten  getLocations in MachineLifecycleEffectorTasks
+     */
+    protected Location findLocation(@Nullable Collection<? extends Location> locations) {
+        if (locations == null || locations.isEmpty()) {
+            locations = this.getLocations();
+        }
+
+        locations = Locations.getLocationsCheckingAncestors(locations, this);
+
+        if (locations.isEmpty())
+            throw new IllegalArgumentException("No locations specified when starting " + this);
+        if (locations.size() != 1 || Iterables.getOnlyElement(locations) == null)
+            throw new IllegalArgumentException("Ambiguous locations detected when starting " + this + ": " + locations);
+        return Iterables.getOnlyElement(locations);
+    }
+
+    //TODO: Probably it would be better an Optional object
+    private CloudFoundryPaasLocation getLocationOrNull() {
+        return Iterables.get(Iterables
```

Comment:
I think it would be more interesting using something like the following:
`````
private Optional<CloudFoundryPaasLocation> tryLocation() {
        return Optional.fromNullable(Iterables.get(Iterables
                .filter(getLocations(), CloudFoundryPaasLocation.class), 0, null));
    }
```` 
in order to return a `Optional<CloudFoundryPaasLocation>`, `tryFilter` would return a `Optional<Location>` instead.